src/formpack/reporting/export.py

```diff
@@ -93,25 +96,53 @@ def __init__(self, formpack, form_versions, lang=UNSPECIFIED_TRANSLATION,
             self._row_cache[section_name] = OrderedDict.fromkeys(fields, '')
             self._empty_row[section_name] = dict(self._row_cache[section_name])
 
-    def parse_submissions(self, submissions):
-        """Return the a generators yielding formatted chunks of the data set"""
-        self.reset()
-        versions = self.versions
-        for entry in submissions:
-            version_id = None
+    def get_version_for_submission(self, submission):
+        """
+        Return the `FormVersion` for this submission, or `None` if none can be
+        found
+        """
+        version_id = None
+        # find the first version_id present in the submission
+        for _key in self.version_id_keys:
+            if _key in submission:
+                version_id = submission.get(_key)
+                break
+        try:
+            return self.versions[version_id]
+        except KeyError:
+            return None
 
-            # find the first version_id present in the submission
-            for _key in self.version_id_keys:
-                if _key in entry:
-                    version_id = entry.get(_key)
-                    break
+    def parse_one_submission(self, submission, version=None):
+        """
+        Parse a single submission and return a formatted 'chunks' structure;
+        see format_one_submission() for details
 
-            try:
-                section = versions[version_id].sections[self.title]
-                submission = FormSubmission(entry)
-                yield self.format_one_submission([submission.data], section)
-            except KeyError:
-                pass
+        :param version: an optional `FormVersion` to use for this submission
+                        instead of inferring the version from the submission
+                        itself
+        """
+        if not version:
+            version = self.get_version_for_submission(submission)
+        if not version:
+            # TODO: somehow include this submission anyway; see
+            # https://github.com/kobotoolbox/formpack/issues/164
+            return None
+        # does this have the same effect as getting the first section?
+        section = version.sections[self.title]
```

Comment:
It's not an attribute of `version.sections`, but rather a key in the dictionary of `sections` held by the `version`. That key [must be there](https://github.com/kobotoolbox/formpack/blob/f04f113f66d28dc11c2526a99f1f45295ee15079/src/formpack/version.py#L117), **but** the code you flagged is confusing and should be changed. Since `FormVersion.sections` [is an `OrderedDict`](https://github.com/kobotoolbox/formpack/blob/f04f113f66d28dc11c2526a99f1f45295ee15079/src/formpack/version.py#L92), and the goal is to get the first section, we could replace this with `version.sections.keys()[0]`. Ultimately, the point of all of this is to have `format_one_submission()` [recurse through all of the sections](https://github.com/kobotoolbox/formpack/blob/f04f113f66d28dc11c2526a99f1f45295ee15079/src/formpack/reporting/export.py#L281-L282).